source/aws_imds_client.c

```diff
@@ -607,15 +615,50 @@ static void s_query_complete(struct imds_user_data *user_data) {
         return;
     }
 
-    /* In this case we fallback to the secure imds flow. */
     if (user_data->status_code == AWS_HTTP_STATUS_CODE_401_UNAUTHORIZED) {
-        s_invalidate_cached_token_safely(user_data);
-        s_reset_scratch_user_data(user_data);
-        aws_retry_token_release(user_data->retry_token);
-        if (s_get_resource_async_with_imds_token(user_data)) {
-            s_user_data_release(user_data);
+        /* In this case we fail the request and switch to secure path for following requests. */
+        if (!user_data->imds_token_required) {
+            if (user_data->is_insecure_fallback) {
+                /* V2 request, but failed to fetch token and fallback failed */
+                AWS_LOGF_ERROR(
+                    AWS_LS_IMDS_CLIENT,
+                    "(id=%p) IMDS client failed to fetch resource through fallback to insecure path. requester %p.",
+                    (void *)user_data->client,
+                    (void *)user_data);
+                user_data->error_code = AWS_AUTH_IMDS_CLIENT_TOKEN_INVALID;
+            } else {
+                /* V1 request, V2 required */
+                AWS_LOGF_ERROR(
+                    AWS_LS_IMDS_CLIENT,
+                    "(id=%p) IMDS client failed to fetch resource without token. requester %p.",
+                    (void *)user_data->client,
+                    (void *)user_data);
+                user_data->error_code = AWS_AUTH_IMDS_CLIENT_TOKEN_REQUIRED;
+            }
+        } else {
+            /* Not expected, the token was told as invalid from server */
+            AWS_LOGF_ERROR(
+                AWS_LS_IMDS_CLIENT,
+                "(id=%p) IMDS client failed to fetch resource. Server response 401 UNAUTHORIZED indicates invalid "
+                "token. requester %p.",
+                (void *)user_data->client,
+                (void *)user_data);
+            user_data->error_code = AWS_AUTH_IMDS_CLIENT_TOKEN_INVALID;
         }
-        return;
+        struct aws_imds_client *client = user_data->client;
+        aws_mutex_lock(&client->token_lock);
+        if (aws_byte_buf_eq(&user_data->imds_token, &client->cached_token)) {
+            /* If the token used matches the cached token, that means the cached token is invalid. */
+            client->token_state = AWS_IMDS_TS_INVALID;
+            AWS_LOGF_DEBUG(
+                AWS_LS_IMDS_CLIENT,
+                "(id=%p) IMDS client's cached token is set to be invalid by requester %p.",
+                (void *)client,
+                (void *)user_data);
+        }
+        /* Flip the flag to let following requests use token as it's required. */
+        client->token_required = true;
+        aws_mutex_unlock(&client->token_lock);
```

Comment:
Discussed offline that we have the following options 

- Maintain the old behavior of attempting at least one v1 request.
-  Deprecate IMDsv1 if the Mountpoint does not require it but they are GA now so this might not be worth a breaking change.
- Perhaps introduce a setter for token required or some sort of environment variable, allowing customers to configure it themselves rather than us changing it in the background without their knowledge and hoping they will retry the request.